### ISSUE TITLE
NeighborSampler producing some incorrect neighbors

### ISSUE THREAD

**Issue by MichaelHillier** [rusty1s/pytorch_geometric#1904]
I have a very simple undirected graph containing 14 nodes that I am testing the NeighborSampler on that is producing some incorrect neighbors for some nodes.

The edges (e.g. adjacency matrix) for the first 4 nodes indexed [0, 1, 2, 3] are:
row = 0, 0, 1, 1, 2, 2, 2, 2, 3, 3, 3
col  = 1, 2, 0, 2, 0, 1, 4, 7, 4, 5, 6

My sampler is:
train_loader = NeighborSampler(data.edge_index, sizes=[-1, -1], batch_size=4)

My sampled neighbors for the batch containing the first 4 nodes is:
row = 0, 0, 1, 1, 2, 2, 2, 2, 3, 3, 3
col  = 1, 2, 0, 2, 0, 1, 4, 5, 4, 6, 7

The edges [2, 5] and [3, 7] from the sampler are incorrect. IF you change the batch_size = 14 (total number of nodes in the graph) the sampler will return the correct neighbors. 

I used pytorch geometric version 1.6.1 and 1.5.0 and both produce the same results.

Have you come across this issue? Is there a way to fix on my end?

**Comment by rusty1s** [comment]
Can you send me the complete adjacency matrix to reproduce this?

**Comment by MichaelHillier** [comment]
Here is a link for a json file that can be loaded using networkx:

https://drive.google.com/file/d/1ne3GOhawPto0b4Ca4PgN4Giz7tHwLBCR/view?usp=sharing

Here is what it looks like along with the adjacency matrix:

![image](https://user-images.githubusercontent.com/29631938/101513912-b4e1eb80-394a-11eb-8e95-2b0d0c8dd925.png)

**Comment by rusty1s** [comment]
I see. You need to keep in mind that the `NeighborSampler` does relabel the node indices (so that the resulting adjacency matrix only contains the sampled nodes), and the `n_id` points to the global node indices. To get the global edge indices, please run:
```python
for batch_size, n_id, adjs in loader:
    for edge_index, _, _ in adjs:
        print(n_id[edge_index])
```

**Comment by MichaelHillier** [comment]
Right. I was not aware that nodes were relabeled for the resulting subgraph. Makes perfect sense. Thanks for your quick reply!

**Comment by MichaelHillier** [comment]
Sorry to reopen, but after experimenting some more I've noticed there is always one odd edge included in layer 2 given the sampled 1-hop for the above graph. Perhaps again I am missing something, but sampling the above graph with **sizes [2, 1]** the resulting output from the loop you provided to print the global edge_indices for the sampling are:

for node batch indexed [0, 1, 2, 3] I get:
2-hop (layer 2) : ` tensor([[0, 0, 1, 3, 4, 5], [1, 2, 0, 4, 3, 6]])`
1-hop (layer 1):  `tensor([[0, 1, 1, 2, 2, 4, 4, 6], [1, 0, 2, 0, 1, 2, 3, 3]]`
odd 2-hop edge [4, 3]. Given the **1-hop** edge [4, 3] appropriate edges for 2-hop neighbors would be either [2, 4], [7, 4], or [6, 4]. 

for node batch indexed [4, 5, 6, 7] I get:
2-hop (layer 2): `tensor([[4, 7, 3, 3, 2, 2, 0], [3, 8, 5, 6, 4, 7, 2]])`
1-hop (layer 1): `tensor([[4, 5, 6, 6, 3, 3, 2, 8], [6, 6, 4, 5, 4, 5, 7, 7]])`
odd 2-hop edge [2, 7] since the edge [7, 4] is not contained in layer 1.

every other batch always has 1 odd 2-hop neighbor.

**Comment by rusty1s** [comment]
Why are the edges `(4, 3)` and `(2, 7)` odd edges? They exist in the graph. Note that we sample nodes for each node involved in the previous iteration. In your first example, that means that we sample edges for nodes `[0, 1, 2, 3, 4, 6]`. This allows the GNN to easily make use of skip-connections, since we have access to intermediate representations of nodes.

**Comment by MichaelHillier** [comment]
In my first example, I would have expected that edges for the nodes `[0, 1, 2, 4, 6]` (not including 3) would be sampled because I thought that the first row of `edge_index` for **layer 1** are the targets for layer 2. But it appears that you sample all the nodes encountered in `edge_index` (both rows: row[0] (source e.g. neighbors), row[1] (target e.g. central nodes)). I don't think this is necessarily a problem, I just wanted to fully understand before I fully scale up my custom GNNs for distributed training on multi-gpus (and later cloud computing). 

Thanks again for your response.

**Comment by rusty1s** [comment]
Yes, that is exactly what happens here.

**Comment by jiaqian** [comment]
> 

Hi, what does sizes  [2,1 ] mean here? Does it mean for the directly connected neighbour, it samples 2, for the neighbour two-step away, it samples 1? Also, is it the same as apply two conv layer with specification 2 for the first layer and 1 for the second layer. I got a bit confused. Thanks in advance.

**Comment by rusty1s** [comment]
Yes, this is exactly what it means.

What do you mean with specification? Is it related to `kernel_size`? It is hard to compare these two TBH. An increase in kernel size means that we aggregate from neighbors more far away. `sizes=[2, 1]` just means that we drop the 1-hop neighborhood (e.g., a `kernel_size=[3, 3]` where we mask certain neighbors).
